gui/tiletypes.lua

```diff
@@ -0,0 +1,1667 @@
+                    y = math.max(math.min(mousePos.y, df.global.world.map.y_count - 1), 0),
+                    z = mousePos.z,
+                }
+                self.last_point = inBoundsMouse
+                self:confirm()
+            end
+        elseif dfhack.maps.isValidTilePos(mousePos.x, mousePos.y, mousePos.z) then
+            self.first_point = self.first_point or mousePos
+        end
+
+        return true
+    end
+
+    -- Switch to the cursor if the cursor was moved (Excluding up and down a Z level)
+    local filteredKeys = utils.clone(keys)
+    filteredKeys["CURSOR_DOWN_Z"] = nil
+    filteredKeys["CURSOR_UP_Z"] = nil
+    self.useCursor = self.useCursor or guidm.getMapKey(filteredKeys)
+        or (self.lastCursorPos and (self.lastCursorPos.x ~= cursorPos.x or self.lastCursorPos.y ~= cursorPos.y))
+    self.lastCursorPos = cursorPos
+
+    return false
+end
+
+function BoxSelection:onRenderFrame(dc, rect)
+    -- Switch to cursor if the mouse is offscreen, or if it hasn't moved
+    self.useCursor = (self.useCursor or (self.lastMousePos and (self.lastMousePos.x < 0 or self.lastMousePos.y < 0)))
+        and self.lastMousePos.x == df.global.gps.precise_mouse_x and self.lastMousePos.y == df.global.gps.precise_mouse_y
+    self.lastMousePos = { x= df.global.gps.precise_mouse_x, y= df.global.gps.precise_mouse_y }
+
+    if self.screen and self.screen.subviews.dimensions_tooltip then
+        self.screen.subviews.dimensions_tooltip.visible = not self.useCursor
+    end
+
+    if self.tile_map then
+        local box = self.box
+
+        if not box then
+            local selectedPos = dfhack.gui.getMousePos(true)
+            if self.useCursor or not selectedPos then
+                selectedPos = copyall(df.global.cursor)
+            end
+
+            if self.flat and self.first_point then
+                selectedPos.z = self.first_point.z
+            end
+
+            local inBoundsMouse = {
+                x = math.max(math.min(selectedPos.x, df.global.world.map.x_count - 1), 0),
+                y = math.max(math.min(selectedPos.y, df.global.world.map.y_count - 1), 0),
+                z = selectedPos.z,
+            }
+
+            box = Box {
+                self.first_point or selectedPos,
+                self.last_point or (self.first_point and inBoundsMouse or selectedPos)
+            }
+        end
+
+        if box then
+            local avoid_rect = type(self.avoid_rect) == "function" and self.avoid_rect() or self.avoid_rect
+            box:draw(self.tile_map, avoid_rect, self.ascii_fill)
+        end
+    end
+
+    -- Don't call super.onRenderFrame, since this widget should not be drawn
+end
+
+function BoxSelection:hideCursor()
+    -- Hide cursor
+    df.global.game.main_interface.main_designation_selected = df.main_designation_type.NONE
+end
+
+--================================--
+--||        SelectDialog        ||--
```

Comment:
It was designed as a more generic version of `gui.materials.MaterialDialog`, since i liked the UI and system, but didn't like how you couldn't set the options or categories yourself, you had to stick to the pre-defined options it gave.
SelectDialog lets you pass in a table with whatever choices and categories you want, even letting you nest categories as deep as you'd like.

Reply:
This is also what led to a lot of the outdated stuff you've commented about, since I copy and pasted `MaterialDialog` and changed it to fit my needs, I brought along a lot of outdated code without realizing.

Reply:
once we get this code finalized, it could maybe move into `gui.dialogs` and `gui.materials.MaterialDialog` could be a subclass of this one